### ISSUE TITLE
webgl: fill does not respect HSB mode

### ISSUE THREAD

**Issue by jcelerier** [processing/p5.js#4974]
#### Most appropriate sub-area of p5.js?

- [ ] Accessibility (Web Accessibility)
- [ ] Build tools and processes
- [x] Color
- [x] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Friendly error system
- [ ] Image
- [ ] IO (Input/Output)
- [ ] Localization
- [ ] Math
- [ ] Unit Testing
- [ ] Typography
- [ ] Utilities
- [x] WebGL
- [ ] Other (specify if possible)

#### Details about the bug:

Given: 
```js
function setup() {
  createCanvas(400, 400, WEBGL);
  background(0);
  colorMode(HSB, 100);
}

function draw() {
  lights();
  
  fill(frameCount % 100, 100, 100);
  stroke(frameCount % 100, 100, 100);
  
  sphere(100);
}
```
the stroke color correctly goes through all hues, but the fill color goes from black to red as if it was still in RGB.

Here is a repro: https://editor.p5js.org/jcelerier/sketches/O3_5dOXA1
Tested on latest edge and chrome on windows.

**Comment by KKVANONYMOUS** [comment]
@jcelerier I guess the `lights()` method creates the illusion that fill is not responding to HSB mode
If you use the following piece of code, you can see the difference.
```
function setup() {
  createCanvas(400, 400, WEBGL);
  background(0);
  colorMode(HSB,100);
}

function draw() {
  // lights();
  
  fill(frameCount % 100, 100, 100);
  stroke(100, frameCount % 100, 100);
  
  sphere(100);
}
```
Note that I have intentionally changed the `stroke` method. Otherwise, we would not be able to differentiate between `fill` and `stroke` methods.

**Comment by jcelerier** [comment]
But what if one wants to use lights() ? e.g. given

```
function setup() {
  createCanvas(400, 400, WEBGL);
  background(0);
  colorMode(HSB,100);
}

function draw() {
  lights();
  noStroke();
  
  fill(frameCount % 100, 100, 100);
  
  sphere(100);
}
``` 

I'd expect the color of the sphere to cycle through the hues, not to go from black to red

**Comment by stalgiag** [comment]
Thanks for the issue @jcelerier ! I think this is due to the fact that `lights()` is then also using HSB. The functions `lights()` calls :
```js
ambientLight(128, 128, 128);
directionalLight(128, 128, 128, 0, 0, -1);
```
which produces the red to black in the example you posted because 128 is a very bright red with `colorMode(HSB, 100)`. A fix for this in the library would be to reset the color mode to default before calling the functions in `lights()` or calling different functions depending on the color mode. In the short term, you can just call your light functions separately without using the `lights()` function. For example, this produces the desired results:

```js
function setup() {
  createCanvas(400, 400, WEBGL);
  background(0);
  colorMode(HSB,100);
}

function draw() {
  noStroke();
  ambientLight(100, 0, 50);
  directionalLight(100, 0, 50, 0, 0, -1);
  fill(frameCount % 100, 100, 100);
  sphere(100);
}
```

**Comment by vulongphan** [comment]
@stalgiag Could I take on this issue?

**Comment by stalgiag** [comment]
@vulongphan definitely!

**Comment by vulongphan** [comment]
Hi @jcelerier, I am pretty new to the code base so could you please guide me in terms of what attributes of p5 class that is responsible for resetting the color mode? 

Thank you!

**Comment by Aloneduckling** [comment]
@stalgiag I just wanted to know that if this would be correct or not 
I haven't tested it yet, but I feel this should work.
```
p5.prototype.lights = function() {
  colorMode(RGB, 255); //default value
  this._assert3d('lights');
  this.ambientLight(128, 128, 128);
  this.directionalLight(128, 128, 128, 0, 0, -1);
  return this;
};
```
Also, I wanted to confirm if we could use `colorMode()` instead of `this.colorMode()` because `colorMode()` is globally accessible.
PS: I thought to add a comment as no one addressed this issue for quite a while.

**Comment by stalgiag** [comment]
Hi @Aloneduckling and @vulongphan ! It might be just a bit more complicated than that because we want this to work in instance mode (so we should use `this.colorMode()`) and we don't want to lose the programmer's previous colorMode so we need to store that beforehand.

Something like this might work but this is untested:
```js
p5.prototype.lights = function() {
  // Make temporary copies of the relevant colorMode properties
  const tempColorMode = this._colorMode;
  // We make a copy in case the colorMode was RGB with different maxes, because the _colorMaxes array might mutate
  const tempColorMaxes = this._colorMaxes[tempColorMode].slice(); 
  colorMode(RGB, 255); //default value
  this._assert3d('lights');
  this.ambientLight(128, 128, 128);
  this.directionalLight(128, 128, 128, 0, 0, -1);
  // Put the temporary values back
  this._colorMode = tempColorMode;
  this._colorMaxes[this._colorMode] = tempColorMaxes
  return this;
};
```
The same thing could be accomplished a bit easier by doing `this.push()` at the beginning and `this.pop()` at the end but it would be significantly less performant because it would be copying all of the rendering properties rather than just the colorMode properties.

Seeing this now -- this issue is a bit more challenging than I originally thought. I am going to remove the level:beginner label but you are still welcome to work on the issue @vulongphan .

**Comment by Aloneduckling** [comment]
@stalgiag Here is the code I wrote. I think it is working; please find the gif below. Is this code ok?
```
p5.prototype.lights = function() {
  const tempColorMode = this._colorMode;
  const tempColorMaxes = this._colorMaxes[tempColorMode].slice();
  this._colorMode = 'rgb';  //default color mode
  this._colorMaxes[this._colorMode] = [255, 255, 255, 255];  //default colorMaxes
  this._assert3d('lights');
  this.ambientLight(128, 128, 128);
  this.directionalLight(128, 128, 128, 0, 0, -1);
  this._colorMode = tempColorMode;
  this._colorMaxes[this._colorMode] = tempColorMaxes;
  return this;
};
```
![asas](https://user-images.githubusercontent.com/54030684/106262836-906b0700-6249-11eb-83ea-eea79e5d0f43.gif)

**Comment by stalgiag** [comment]
Hi @Aloneduckling that is good. I realize that the `colorMode` restoring should probably only happen in instances where `colorMode` isn't set to default. So this should be checked first. `lights()` will often run in `draw()` and it is probably better to avoid the array copy when possible. 

But this should also be for @vulongphan to try first if they want. Issue assignments are first-come, first-serve so no one else can open a PR but @vulongphan unless they say they no longer want to work on it. Your help in working through the solution to this has been useful and I am sure you have learned a lot about contributing by working on it. Feel free to ask to be assigned to another issue while this one continues to be worked on by @vulongphan until further notice.

**Comment by Aloneduckling** [comment]
@stalgiag Great! @vulongphan can work on this. I am here to contribute and learn, I will be hunting other issues.

PS. The documentation is very good and is helping me a lot.

**Comment by vulongphan** [comment]
Thank you @stalgiag and @Aloneduckling for your help in assisting me to tackle this issue. My code is built upon the solution presented by @Aloneduckling but I carry out the check for whether the `colorMode` is already in default mode
```
p5.prototype.lights = function () {
 if (this._colorMode !== 'rgb') { // only restore the colorMode to default if it is not in default already
    const tempColorMode = this._colorMode;
    const tempColorMaxes = this._colorMaxes[tempColorMode].slice();
    this._colorMode = 'rgb';  //default color mode
    this._colorMaxes[this._colorMode] = [255, 255, 255, 255];  //default colorMaxes
    this._assert3d('lights');
    this.ambientLight(128, 128, 128);
    this.directionalLight(128, 128, 128, 0, 0, -1);
    this._colorMode = tempColorMode;
    this._colorMaxes[this._colorMode] = tempColorMaxes;
  }
  else {
    this._assert3d('lights');
    this.ambientLight(128, 128, 128);
    this.directionalLight(128, 128, 128, 0, 0, -1);
  }
  return this;
};
```
This gives the desired result.
@stalgiag  if you think this is okay, I will open a PR. 
Thank you!

**Comment by stalgiag** [comment]
Yeah definitely feel free to open a PR @vulongphan . Good work team!

**Comment by stalgiag** [comment]
@all-contributors please add @jcelerier for bug reports

This issue was resolved by #5049
